### PR TITLE
[ONNX] Enable constant folding for Shape

### PR DESCRIPTION
**Pull request by KsenijaS**
Enabled constant folding for onnx::Shape

### CODE REVIEW

File: test/onnx/test_utility_funs.py [hunk 1]
```diff
@@ -309,6 +309,27 @@ def forward(self, x):
...
         assert len(list(graph.nodes())) == 1
 
+    def test_constant_fold_shape(self):
+        class ShapeModule(torch.nn.Module):
+            def __init__(self):
+                super(ShapeModule, self).__init__()
+                self.register_buffer("weight", torch.ones(5))
+
+            def forward(self, x):
+                shape = torch.tensor(list(self.weight.shape))
```
Comment (neginraoof) [inline test/onnx/test_utility_funs.py]: nit: list() is not necessary here. Could be: shape = torch.tensor(self.weight.shape)

**Review by neginraoof — APPROVED** [review]

File: test/onnx/test_utility_funs.py [hunk 2]
```diff
@@ -309,6 +309,27 @@ def forward(self, x):
...
+                return x * shape
```
Comment (neginraoof) [inline test/onnx/test_utility_funs.py]: Actually, I don't think shape node appears in this model even without constant folding. Could you please try constant_folding=False and try this?
You can have the model defined as:
```
            def forward(self, x):
                shape = self.weight.shape[0]
                return x * shape
```
And set len(list(graph.nodes())) == 4 maybe.

**Review by neginraoof — CHANGES_REQUESTED** [review]

**Comment by neginraoof** [comment]
Also, please check the failures. Looks like test test/onnx/test_operators.py TestOperators.test_expand is failing.
Export of expand op in opset 9 uses ones_like op which uses shape op. This could be expected, please verify and update the expand expected graph file.

**Review by neginraoof — COMMENTED** [review]
Thanks for the fixes. Please check if there's an onnxruntime test for shape and add that if not.

**Review by neginraoof — APPROVED** [review]

**Comment by KsenijaS** [comment]
@houseroad can you merge this PR?

**Comment by neginraoof** [comment]
@KsenijaS could you rebase this PR please?

**Comment by KsenijaS** [comment]
@houseroad can you please review my PR, it seems that none of the failure are caused by my changes

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 3]
```diff
@@ -295,7 +299,7 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
         inputTensorValues[0], p, node->is(attr::axes), node->i(attr::keepdims));
     return c10::optional<at::Tensor>(updated_val);
   } else if (node->kind() == onnx::Gather) {
-    assert(inputTensorValues.size() == 1);
+    assert(inputTensorValues.size() == 2);
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp:302]: why change this?

**Review by houseroad — COMMENTED** [review]
not sure why we need to change onnx::gather part

File: torch/csrc/jit/passes/onnx/constant_fold.cpp (hunk 3 above)
Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/constant_fold.cpp:302]: gather has 2 inputs, that's why my tests were failing

Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/constant_fold.cpp:302]: This was a bug in gather implementation for constant folding. This node is required to have 2 inputs always (Note that this assertion as not blocking tests on CI, and was caught locally).

**Comment by KsenijaS** [comment]
@houseroad can we merge this PR?

File: torch/csrc/jit/passes/onnx/constant_fold.cpp (hunk 3 above)
Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/constant_fold.cpp:302]: @houseroad if you have no further comments, can this PR be merged?

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 4]
```diff
@@ -282,6 +282,10 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
       }
     }
     return c10::optional<at::Tensor>(at::reshape(updated_val, shape));
+  } else if (node->kind() == onnx::Shape) {
+    assert(inputTensorValues.size() == 1);
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp:286]: let's update assert to TORCH_INTERNAL_ASSERT

**Review by BowenBao — APPROVED** [review]
LGTM, please update assert to TORCH_INTERNAL_ASSERT

**Review by houseroad — APPROVED** [review]
Thanks

**Pull request closed without merging**